funsor/einsum/util.py

```diff
@@ -56,3 +60,48 @@ def __call__(self, x, y, axes=2):
         # form full string and contract!
         einsum_str = "{},{}->{}".format(*map("".join, (x_ix, y_ix, out_ix)))
         return self.einsum(einsum_str, x, y)
+
+
+def rename_equation(equation, *operands):
+    """
+    Renames symbols in an einsum/ubersum equation to match the
+    ``.pyro_dims`` attributes of packed ``operands``.
+    """
+    inputs, outputs = equation.split('->')
+    inputs = inputs.split(',')
+    assert len(inputs) == len(operands)
+    rename = {old: new
+              for input_, operand in zip(inputs, operands)
+              for old, new in zip(input_, operand._pyro_dims)}
+    return ''.join(rename.get(s, s) for s in equation)
+
+
+class Array(np.ndarray):
+    pass
+
+
+def broadcast_all(*values, **kwargs):
+    """
+    Packed broadcasting of multiple tensors.
+    """
+    dims = kwargs.get('dims')
+    sizes = {dim: size for value in values for dim, size in zip(value._pyro_dims, value.shape)}
+    if dims is None:
+        dims = ''.join(sorted(sizes))
+    else:
+        assert set(dims) == set(sizes)
+    shape = tuple(sizes[dim] for dim in dims)
+    values = list(values)
+    for i, x in enumerate(values):
+        old_dims = x._pyro_dims
+        if old_dims != dims:
+            x = ops.permute(x, tuple(old_dims.index(dim) for dim in dims if dim in old_dims))
+            x = x.reshape(tuple(sizes[dim] if dim in old_dims else 1 for dim in dims))
+            x = ops.expand(x, shape)
+            # workaround: ndarray does not allow setting attribute "_pyro_dims"
```

Comment:
I think this is an important point: we can't set new attributes to ndarray, DeviceArray directly.